communication/src/server.cpp

```
@@ -0,0 +1,271 @@
+    socketInterface->close(*it);
+    if (it != sockets.end())
+        sockets.erase(it);
+    }
+    {
+        std::lock_guard<std::mutex> lock(IDMapMutex);
+        clientIDMap.erase(clientSocket);
+    }
+}
+
+// Implementation according to the CAN BUS
+bool Server::isValidId(uint32_t id)
+{
+    if(id==0)
+        return false;
+        
+    //implement id exist already
+    return true;
+}
+// Returns the sockets ID
+int Server::getClientSocketByID(uint32_t destID)
+{
+    std::lock_guard<std::mutex> lock(IDMapMutex);
+    for (const auto &client : clientIDMap)
+        if (client.second == destID)
+            return client.first;
+
+    return -1;
+}
+
+// Sends the message to destination
+ErrorCode Server::sendDestination(const Packet &packet)
+{
+    int targetSocket = getClientSocketByID(packet.header.DestID);
+    if (targetSocket == -1)
+        return ErrorCode::INVALID_CLIENT_ID;
+    
+    ssize_t bytesSent = socketInterface->send(targetSocket, &packet, sizeof(Packet), 0);
+    if (!bytesSent)
+        return ErrorCode::SEND_FAILED;
+
+    if (bytesSent<0){
+        //closeConnection();
+        return ErrorCode::CONNECTION_FAILED;
+    }
+    return ErrorCode::SUCCESS;
+}
+
+// Sends the message to all connected processes - broadcast
+ErrorCode Server::sendBroadcast(const Packet &packet)
+{
+    std::lock_guard<std::mutex> lock(socketMutex);
+    for (int sock : sockets) {
+        ssize_t bytesSent = socketInterface->send(sock, &packet, sizeof(Packet), 0);
+        if (bytesSent < sizeof(Packet))
+            return ErrorCode::SEND_FAILED;
+        if (bytesSent<0){
+        //closeConnection();
+        return ErrorCode::CONNECTION_FAILED;
+        }
+    }
+
+    return ErrorCode::SUCCESS;
+}
+
+// Sets the server's port number, throws an exception if the port is invalid.
+void Server::setPort(int port) {
+    if (port <= 0 || port > 65535)
+        throw std::invalid_argument("Invalid port number: Port must be between 1 and 65535.");

```

should return an error code?